src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -21,489 +21,211 @@
-    recordNumHistory_.push_back(recordNum);
-    if (patternNZHistory_.size() > maxSteps_) {
-      patternNZHistory_.pop_front();
-      recordNumHistory_.pop_front();
-    }
-  }
-
-  // if input pattern has greater index than previously seen, update
-  // maxInputIdx and augment weight matrix with zero padding
-  if (patternNZ.size() > 0) {
-    const UInt maxInputIdx = *max_element(patternNZ.begin(), patternNZ.end());
-    if (maxInputIdx > maxInputIdx_) {
-      maxInputIdx_ = maxInputIdx;
-    }
-  }
-
-  // if in inference mode, compute likelihood and update return value
-  if (infer) {
-    infer_(patternNZ, actValueList, result);
-  }
-
-  // update weights if in learning mode
-  if (learn) {
-    for (size_t categoryI = 0; categoryI < bucketIdxList.size(); categoryI++) {
-      const UInt bucketIdx = bucketIdxList[categoryI];
-      const Real64 actValue = actValueList[categoryI];
-      // if bucket is greater, update maxBucketIdx_ and augment weight
-      // matrix with zero-padding
-      if (bucketIdx > maxBucketIdx_) {
-        maxBucketIdx_ = bucketIdx;
-      }
-
-      // update rolling averages of bucket values
-      while (actualValues_.size() <= maxBucketIdx_) {
-        actualValues_.push_back(0.0);
-        actualValuesSet_.push_back(false);
-      }
-      if (!actualValuesSet_[bucketIdx] || category) {
-        actualValues_[bucketIdx] = actValue;
-        actualValuesSet_[bucketIdx] = true;
-      } else {
-        actualValues_[bucketIdx] =
-            ((1.0 - actValueAlpha_) * actualValues_[bucketIdx]) +
-            (actValueAlpha_ * actValue);
-      }
-    }
-
-    // compute errors and update weights
-    auto patternIteration = patternNZHistory_.begin();
-    for (auto learnRecord = recordNumHistory_.begin();
-         learnRecord != recordNumHistory_.end();
-         learnRecord++, patternIteration++) {
-      const vector<UInt> learnPatternNZ = *patternIteration;
-      const UInt nSteps = recordNum - *learnRecord;
-
-      // update weights
-      if (binary_search(steps_.begin(), steps_.end(), nSteps)) {
-        const vector<Real64> error = calculateError_(bucketIdxList, learnPatternNZ, nSteps);
-        Matrix& w = weightMatrix_.at(nSteps);
-	NTA_ASSERT(alpha_ > 0.0);
-        for (const auto& bit : learnPatternNZ) {
-          for(size_t i = 0; i < error.size(); i++) {
-            const auto val = get_(w, bit, i) + alpha_ * error[i];
-	    if(val == 0) continue;
-	    w[bit][i] = val;
-          }
-        }
-      }
-    }
-  }
+PDF Classifier::infer(const SDR & pattern)
+{
+  // Check input dimensions, or if this is the first time the Classifier has
+  // been used then initialize it with the given SDR's dimensions.
+  if( dimensions_.empty() ) {
+    dimensions_ = pattern.dimensions;
+    while( weights_.size() < pattern.size ) {
+      weights_.push_back( vector<Real>( numCategories_, 0.0f ));
+    }
+  } else if( pattern.dimensions != dimensions_ ) {
```

Comment:
I'd like for this check to always run.  Otherwise its liable to segfault if the user uses the wrong SDR.

Reply:
the NTA_CHECK, but not a high priority, so as you want

Reply:
I added the extra verbose output was in response to a specific and frustrating error.  The following is incorrect but will compile:
```
Classifier C();
vector<UInt> sparse({ 4,5,6 });
C.infer( sparse );
```
Instead of failing because the input `sparse` is not an SDR, the C++ compiler creates an SDR with dimensions [4,5,6].

Reply:
ok, that is a really cunning of the compiler! Good catch